File: webapp/src/shared/openai-models.js
export const DEFAULT_MODEL = "gpt-4.1-mini";

export const MODEL_SUGGESTIONS = [
  "gpt-5.4",
  "gpt-5.4-mini",
  "gpt-5.4-nano",
  "gpt-5.4-pro",
  "gpt-5.2",
  "gpt-5.2-pro",
  "gpt-5.1",
  "gpt-5",
  "gpt-5-mini",
  "gpt-5-nano",
  "gpt-5-pro",
  "gpt-4.1",
  "gpt-4.1-mini",
  "gpt-4.1-nano",
  "gpt-4o",
  "gpt-4o-mini",
  "o4-mini",
  "o3"
];

function normalizeLower(value) {
  return String(value || "").trim().toLowerCase();
}

export function normalizeModelName(model) {
  const trimmed = String(model || "").trim();
  return trimmed || DEFAULT_MODEL;
}

export function getReasoningProfile(model) {
  const normalized = normalizeLower(model);

  if (!normalized) {
    return {
      supported: true,
      options: ["none", "minimal", "low", "medium", "high", "xhigh"],
      defaultValue: "medium"
    };
  }

  if (normalized.includes("pro")) {
    return {
      supported: true,
      options: ["high"],
      defaultValue: "high"
    };
  }

  if (normalized.startsWith("gpt-5.1")) {
    return {
      supported: true,
      options: ["none", "low", "medium", "high"],
      defaultValue: "none"
    };
  }

  if (
    normalized.startsWith("gpt-5.2") ||
    normalized.startsWith("gpt-5.3") ||
    normalized.startsWith("gpt-5.4") ||
    normalized.startsWith("gpt-5.5") ||
    normalized.startsWith("gpt-6")
  ) {
    return {
      supported: true,
      options: ["none", "minimal", "low", "medium", "high", "xhigh"],
      defaultValue: "medium"
    };
  }

  if (normalized.startsWith("gpt-5") || normalized.startsWith("o")) {
    return {
      supported: true,
      options: ["minimal", "low", "medium", "high"],
      defaultValue: "medium"
    };
  }

  return {
    supported: false,
    options: [],
    defaultValue: null
  };
}

export function normalizeReasoningEffort(model, reasoningEffort) {
  const profile = getReasoningProfile(model);

  if (!profile.supported) {
    return null;
  }

  const normalized = normalizeLower(reasoningEffort);

  if (normalized && profile.options.includes(normalized)) {
    return normalized;
  }

  return profile.defaultValue;
}

File: goserver/internal/shared/openai_models.go
package shared

import "strings"

const DefaultModel = "gpt-4.1-mini"

type ReasoningProfile struct {
	Supported    bool
	Options      []string
	DefaultValue string
}

func NormalizeModelName(model string) string {
	trimmed := strings.TrimSpace(model)
	if trimmed == "" {
		return DefaultModel
	}

	return trimmed
}

func GetReasoningProfile(model string) ReasoningProfile {
	normalized := normalizeLower(model)

	if normalized == "" {
		return ReasoningProfile{
			Supported:    true,
			Options:      []string{"none", "minimal", "low", "medium", "high", "xhigh"},
			DefaultValue: "medium",
		}
	}

	if strings.Contains(normalized, "pro") {
		return ReasoningProfile{
			Supported:    true,
			Options:      []string{"high"},
			DefaultValue: "high",
		}
	}

	if strings.HasPrefix(normalized, "gpt-5.1") {
		return ReasoningProfile{
			Supported:    true,
			Options:      []string{"none", "low", "medium", "high"},
			DefaultValue: "none",
		}
	}

	if strings.HasPrefix(normalized, "gpt-5.2") ||
		strings.HasPrefix(normalized, "gpt-5.3") ||
		strings.HasPrefix(normalized, "gpt-5.4") ||
		strings.HasPrefix(normalized, "gpt-5.5") ||
		strings.HasPrefix(normalized, "gpt-6") {
		return ReasoningProfile{
			Supported:    true,
			Options:      []string{"none", "minimal", "low", "medium", "high", "xhigh"},
			DefaultValue: "medium",
		}
	}

	if strings.HasPrefix(normalized, "gpt-5") || strings.HasPrefix(normalized, "o") {
		return ReasoningProfile{
			Supported:    true,
			Options:      []string{"minimal", "low", "medium", "high"},
			DefaultValue: "medium",
		}
	}

	return ReasoningProfile{}
}

func NormalizeReasoningEffort(model, reasoningEffort string) *string {
	profile := GetReasoningProfile(model)
	if !profile.Supported {
		return nil
	}

	normalized := normalizeLower(reasoningEffort)
	for _, option := range profile.Options {
		if normalized == option {
			return StringPtr(option)
		}
	}

	return StringPtr(profile.DefaultValue)
}

func normalizeLower(value string) string {
	return strings.ToLower(strings.TrimSpace(value))
}
